services/web-server/src/login/strategies/github.js

```diff
@@ -39,18 +42,22 @@ module.exports = class Github {
     return user;
   }
 
-  async userFromToken(accessToken) {
-    const githubClient = new GithubClient();
-    const [githubErr, githubUser] = await tryCatch(githubClient.userFromToken(accessToken));
+  async userFromToken(token) {
+    const [jwtError, jwtResponse] = await tryCatch(
+      jwt.verify({ publicKey: this.jwt.publicKey, token })
+    );
```

Comment:
I think the strategies should not be responsible for validating the JWT.  In particular, one thing that's missing here is validating that the JWT is for this identity provider.  Here's a threat scenario: a deployment has two strategies configured, A and B.  It turns out that `hassan` isn't taken as a username on B, so I claim that username and sign in, getting a JWT with `{sub: "hassan"}`.  Then I call `getCredentials(identityProvider: "A", taskclusterToken: <token>)` and magically I'm logged in as you on A.

So, I'd suggest two changes:
 * validate the JWT in `getCredentials`
 * put the *identity* (including the provider) in the `sub` claim, and use that to determine which strategy to call.

Then this `userFromToken` method can become userFromIdentity` instead (since it just needs the identity from the `sub` claim to do its thing), in both strategies.

Reply:
> put the identity (including the provider) in the sub claim, and use that to determine which strategy to call.

I think this is already being done in https://github.com/taskcluster/taskcluster/pull/826/files#diff-5837766ca22fc9915952fbe10ebaa9f9R231. I should probably just rename this field to sub to make it more explicit that we're setting the sub field to the identity.